test/e2e/nsx_subnet_test.go

```diff
@@ -750,3 +793,547 @@ func fetchSubnetPortBySubnetPortUID(t *testing.T, subnetPortUID string) *model.V
+	var actionMsg string
+	switch {
+	case expectedCount == 0:
+		actionMsg = "to be removed"
+	case expectedCount == 1:
+		actionMsg = "to be created"
+	default:
+		actionMsg = fmt.Sprintf("%d to be created", expectedCount)
+	}
+
+	log.Info(fmt.Sprintf("Waiting for shared subnet(s) %s", actionMsg), "namespace", namespace)
+	var sharedSubnets []*v1alpha1.Subnet
+
+	err := wait.PollUntilContextTimeout(context.TODO(), 1*time.Second, 100*time.Second, false, func(ctx context.Context) (bool, error) {
+		// Get all subnets from the namespace
+		subnetList, err := listNamespaceSubnets(ctx, namespace)
+		if err != nil {
+			return false, err
+		}
+
+		// Find and collect shared subnets
+		sharedSubnets = []*v1alpha1.Subnet{}
+		log.Info("Checking subnets in namespace", "namespace", namespace, "count", len(subnetList.Items))
+		for i := range subnetList.Items {
+			subnet := &subnetList.Items[i]
+			log.Info("Examining subnet", "namespace", namespace, "name", subnet.Name, "shared", subnet.Status.Shared)
+			if subnet.Status.Shared {
+				sharedSubnets = append(sharedSubnets, subnet)
+			}
+		}
+
+		// Check if we have the expected number of shared subnets
+		if len(sharedSubnets) != expectedCount {
+			log.Info("Waiting for expected number of shared subnets",
+				"namespace", namespace,
+				"currentCount", len(sharedSubnets),
+				"expectedCount", expectedCount)
+			return false, nil
+		}
+
+		// If expectedCount is 0, we're done
+		if expectedCount == 0 {
+			return true, nil
+		}
+
+		// Verify name prefixes if provided and if we expect subnets
+		if len(namePrefixes) > 0 {
+			for _, subnet := range sharedSubnets {
+				hasExpectedPrefix := false
+				for _, prefix := range namePrefixes {
+					if strings.HasPrefix(subnet.Name, prefix) {
+						hasExpectedPrefix = true
+						break
+					}
+				}
+				if !hasExpectedPrefix {
+					log.Info("Shared subnet has unexpected name prefix",
+						"namespace", namespace,
+						"name", subnet.Name,
+						"expectedPrefixes", namePrefixes)
+					return false, nil
+				}
+			}
+		}
+
+		return true, nil
+	})
+
+	// Handle assertions based on expected count
+	if expectedCount == 0 {
+		require.NoError(t, err, "Shared subnet should be removed from %s", namespace)
+	} else {
+		require.NoError(t, err, "Failed to find %d shared subnet(s) in %s", expectedCount, namespace)
+		require.Equal(t, expectedCount, len(sharedSubnets),
+			"Expected %d shared subnet(s) in %s, got %d", expectedCount, namespace, len(sharedSubnets))
+
+		// Additional validation for single subnet case
+		if expectedCount == 1 && len(namePrefixes) > 0 {
+			require.NotNil(t, sharedSubnets[0], "Shared subnet in %s should not be nil", namespace)
+			require.True(t, strings.HasPrefix(sharedSubnets[0].Name, namePrefixes[0]),
+				"Shared subnet name should be prefixed with '%s', got: %s", namePrefixes[0], sharedSubnets[0].Name)
+		}
```

Comment:
Do we need to check it again here as I think it is check at line 966-983?
 https://github.com/vmware-tanzu/nsx-operator/blob/cf4986db7c2bfa9726b7f70a0ced96cefc15f59b/test/e2e/nsx_subnet_test.go#L966-L983